Make suppressing and storing Twisted logs optional

jml opened the pull request:
Fixes https://bugs.launchpad.net/testtools/+bug/1515362 and https://bugs.launchpad.net/testtools/+bug/942785. Depends on #171.

Adds two parameters to `AsynchronousDeferredRunTest` that allow callers to decide whether we suppress Twisted logs and whether we capture Twisted logs. Also exposes `CaptureTwistedLogs` so that users who want to disable the automatic capturing can use the fixture to do it themselves, thus allowing post-processing of logs information.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/testing-cabal/testtools/172)
<!-- Reviewable:end -->

rbtcollins commented:
I wonder if you might rebase this slightly - you have a lot of exploratory and feature-branch-merge noise in the history, and it makes reviewing whats happened in e.g. a release quite a bit harder :(.  If it was bzr or loom, there'd be tooling to help :)

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/testing-cabal/testtools/172#-:-K9Bjih7go-WfiwCAvLJ:708881559)*
<!-- Sent from Reviewable.io -->

jml commented:
Rebased. I can also split into two commits (one for exposing `CaptureTwistedLogs`, one for the rest) if you'd like.

Took me a little while fiddling w/ reviewable.io to make it show the right diffs.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/testing-cabal/testtools/172#-:-K9C8D8JmDlc91deosje:1822566517)*
<!-- Sent from Reviewable.io -->

thomir commented on NEWS:
```diff
@@ -9,6 +9,14 @@ Next
+  logging behaviour.  (Jonathan Lange, #942785)
+
+* New fixture ``CaptureTwistedLogs`` that can be used with
+  ``AsynchronousDeferredRunTest`` to attach a detail containing everything
```
If I'm not mistaken, CaptureTwistedLogs can be used with any test runner? If so, suggest removing 'with ``Async...``' as it sort of implies that it's only useful in that case. But perhaps I've misunderstood....

rbtcollins replied:
Its a twisted specifical log capturerer, but not specific to testtools; thats true. OTOH testtools is currently the only twisted and details supporting runner; I think being overly general will make the changelog harder to understand.

Which reminds me: docs please; shouldn't this be in the for-test-authors docs?

jml commented:
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[NEWS, line 17 \[r2\]](https://reviewable.io:443/reviews/testing-cabal/testtools/172#-K4wZdHtocTumUdcLNTv-r2-17:-K9YqAf2G-0d3Oze0ZoB:34595947) ([raw file](https://github.com/testing-cabal/testtools/blob/1e2aa0d89f84718e6bcc959e68c33c54d69afbdf/NEWS#L17)):*
It should be in `twisted-support.rst`. OK for me to land after adding those, or should I get another round of review?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/testing-cabal/testtools/172)*
<!-- Sent from Reviewable.io -->

rbtcollins commented:
Land away imo
On 3 Feb 2016 8:20 AM, "Jonathan Lange" <notifications@github.com> wrote:

> Review status: 0 of 4 files reviewed at latest revision, 1 unresolved
> discussion.
> ------------------------------
>
> *NEWS, line 17 [r2]
> <https://reviewable.io:443/reviews/testing-cabal/testtools/172#-K4wZdHtocTumUdcLNTv-r2-17:-K9YqAf2G-0d3Oze0ZoB:34595947>
> (raw file
> <https://github.com/testing-cabal/testtools/blob/1e2aa0d89f84718e6bcc959e68c33c54d69afbdf/NEWS#L17>):*
> It should be in twisted-support.rst. OK for me to land after adding
> those, or should I get another round of review?
> ------------------------------
>
> *Comments from the review on Reviewable.io
> <https://reviewable.io:443/reviews/testing-cabal/testtools/172>*
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/testing-cabal/testtools/pull/172#issuecomment-178826725>
> .
>

jml commented:
Merged in  302751e
